src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -916,810 +1004,330 @@ private JCExpression makeIndyCall(DiagnosticPosition pos, Type site, Name bsmNam
+            return names.fromString(result);
         }
 
         /**
-         * This class is used to store important information regarding translation of
-         * lambda expression/method references (see subclasses).
+         * Translate a symbol of a given kind into something suitable for the
+         * synthetic lambda body
          */
-        abstract class TranslationContext<T extends JCFunctionalExpression> {
-
-            /** the underlying (untranslated) tree */
-            final T tree;
-
-            /** points to the adjusted enclosing scope in which this lambda/mref expression occurs */
-            final Symbol owner;
-
-            /** the depth of this lambda expression in the frame stack */
-            final int depth;
-
-            /** the enclosing translation context (set for nested lambdas/mref) */
-            final TranslationContext<?> prev;
-
-            /** list of methods to be bridged by the meta-factory */
-            final List<Symbol> bridges;
-
-            TranslationContext(T tree) {
-                this.tree = tree;
-                this.owner = owner(true);
-                this.depth = frameStack.size() - 1;
-                this.prev = context();
-                ClassSymbol csym =
-                        types.makeFunctionalInterfaceClass(attrEnv, names.empty, tree.target, ABSTRACT | INTERFACE);
-                this.bridges = types.functionalInterfaceBridges(csym);
+        VarSymbol translate(final VarSymbol sym, LambdaSymbolKind skind) {
+            VarSymbol ret;
+            switch (skind) {
+                case CAPTURED_VAR:
+                    Name name = (sym.flags() & LOCAL_CAPTURE_FIELD) != 0 ?
+                            sym.baseSymbol().name : sym.name;
+                    ret = new VarSymbol(SYNTHETIC | FINAL | PARAMETER, name, types.erasure(sym.type), translatedSym);
+                    break;
+                case LOCAL_VAR:
+                    ret = new VarSymbol(sym.flags() & FINAL, sym.name, sym.type, translatedSym);
+                    ret.pos = sym.pos;
+                    // If sym.data == ElementKind.EXCEPTION_PARAMETER,
+                    // set ret.data = ElementKind.EXCEPTION_PARAMETER too.
+                    // Because method com.sun.tools.javac.jvm.Code.fillExceptionParameterPositions and
+                    // com.sun.tools.javac.jvm.Code.fillLocalVarPosition would use it.
+                    // See JDK-8257740 for more information.
+                    if (sym.isExceptionParameter()) {
+                        ret.setData(ElementKind.EXCEPTION_PARAMETER);
+                    }
+                    break;
+                case PARAM:
+                    ret = new VarSymbol((sym.flags() & FINAL) | PARAMETER, sym.name, types.erasure(sym.type), translatedSym);
+                    ret.pos = sym.pos;
+                    // Set ret.data. Same as case LOCAL_VAR above.
+                    if (sym.isExceptionParameter()) {
```

Comment:
do we really need to set this exception parameter thing for a method param?

Reply:
I'm not 100% sure, but eyeballing the code, I see that `sym.isExceptionParameter` is used in the backend (e.g. `Code`) so I'd suspect that a failure in doing this would lead to issues when generating bytecode, which I suspect is why this code was added in the first place. But, while that seems useful for a local variable declared inside a lambda, I don't see how a lambda parameter can itself be an exception parameter...